src/py_backed.rs

```diff
@@ -0,0 +1,117 @@
+use std::ops::Deref;
+
+use crate::{
+    types::{
+        any::PyAnyMethods, bytearray::PyByteArrayMethods, bytes::PyBytesMethods,
+        string::PyStringMethods, PyByteArray, PyBytes, PyString,
+    },
+    Bound, DowncastError, FromPyObject, Py, PyAny, PyResult,
+};
+
+/// A wrapper around `str` where the storage is owned by a Python `bytes` or `str` object.
+///
+/// This type gives access to the underlying data via a `Deref` implementation.
+pub struct PyBackedStr {
+    #[allow(dead_code)]
+    storage: PyBackedStrStorage,
+    data: *const u8,
+    length: usize,
+}
+
+#[allow(dead_code)]
+enum PyBackedStrStorage {
+    String(Py<PyString>),
+    Bytes(Py<PyBytes>),
+}
+
+impl Deref for PyBackedStr {
+    type Target = str;
+    fn deref(&self) -> &str {
+        unsafe {
+            // Safety: `data` is a pointer to the start of a valid UTF-8 string of length `length`.
+            std::str::from_utf8_unchecked(std::slice::from_raw_parts(self.data, self.length))
+        }
+    }
+}
+
+impl FromPyObject<'_> for PyBackedStr {
+    fn extract_bound(obj: &Bound<'_, PyAny>) -> PyResult<Self> {
+        let py_string = obj.downcast::<PyString>()?;
+        #[cfg(any(Py_3_10, not(Py_LIMITED_API)))]
+        {
+            let s = py_string.to_str()?;
+            let data = s.as_ptr();
+            let length = s.len();
+            Ok(Self {
+                storage: PyBackedStrStorage::String(py_string.to_owned().unbind()),
+                data,
+                length,
+            })
+        }
+        #[cfg(not(any(Py_3_10, not(Py_LIMITED_API))))]
+        {
+            let bytes = string.encode_utf8()?;
+            let b = bytes.as_bytes();
+            let data = b.as_ptr();
+            let length = b.len();
+            Ok(Self {
+                storage: PyBackedStrStorage::Bytes(bytes.unbind()),
+                data,
+                length,
+            })
+        }
+    }
+}
+
+/// A wrapper around `[u8]` where the storage is either owned by a Python `bytes` object, or a Rust `Vec<u8>`.
+///
+/// This type gives access to the underlying data via a `Deref` implementation.
+pub struct PyBackedBytes {
+    #[allow(dead_code)] // only held so that the storage is not dropped
+    storage: PyBackedBytesStorage,
+    data: *const u8,
```

Comment:
Could these data pointers be `NonNull`? It would provide one more niche for the compiler to use elsewhere.

I also wonder whether we could store `*const [u8]` or even `NonNull<*const [u8]>` or whether pointer metadata handling is not up it on our MSRV yet.

Reply:
`NonNull<[u8]>` is a really elegant idea, let's see what MSRV says...